src/msg/producer/writer/consumer_writer.go

```diff
@@ -171,18 +194,34 @@ func (w *consumerWriterImpl) Address() string {
 
 // Write should fail fast so that the write could be tried on other
 // consumer writers that are sharing the message queue.
-func (w *consumerWriterImpl) Write(b []byte) error {
-	if !w.validConn.Load() {
+func (w *consumerWriterImpl) Write(connIndex int, b []byte) error {
+	w.writeState.RLock()
+	if !w.writeState.validConns || len(w.writeState.conns) == 0 {
+		w.writeState.RUnlock()
 		w.m.writeInvalidConn.Inc(1)
 		return errInvalidConnection
 	}
-	w.writeLock.Lock()
-	_, err := w.rw.Write(b)
-	w.writeLock.Unlock()
+	if connIndex < 0 || connIndex >= len(w.writeState.conns) {
+		w.writeState.RUnlock()
+		return fmt.Errorf("connection index out of range: %d", connIndex)
+	}
+
+	writeConn := w.writeState.conns[connIndex]
+
+	// Make sure only writer to this connection.
+	writeConn.writeLock.Lock()
+	_, err := writeConn.rw.Write(b)
+	writeConn.writeLock.Unlock()
+
+	// Hold onto the write state lock until done, since flushing and
+	// closing connections are done by acquiring the write state lock.
```

Comment:
nit: maybe put this in a function that runs with read lock? Then won't have to do all of these `RUnlock()`s which may cause problems in the future?